### PR TITLE
feat: implement create_liquidation_order function

### PR DESCRIPTION
**Pull request by lambda-0x**
# Pull Request type

Please add the labels corresponding to the type of changes your PR introduces:

- Feature

## What is the current behavior?

Resolves: #263 

## What is the new behavior?

- [x] implements the relevant function
- [ ] see what kind of tests can be added
-

## Does this introduce a breaking change?

No

### CODE REVIEW

**Comment by lambda-0x** [comment]
hey guys, i need some help with failing tests. I am not sure why its failing.

anyway to debug this further?

cc: @Sk8erboi84 Sk8erboi84

**Comment by Sk8erboi84** [comment]
> hey guys, i need some help with failing tests. I am not sure why its failing.
> 
> anyway to debug this further?
> 
> cc: @Sk8erboi84 Sk8erboi84

The issue is in ```test_exec_liquidation_true``` test function, which calls a function which invokes your ```create_liquidation_order``` function. To fix this issue, you need to modify that test because it is currently showing that the position is not found.

**Comment by lambda-0x** [comment]
okay, so the issue is test need to create a position first before trying to liquidate it.

**Comment by lambda-0x** [comment]
i was able to fix the old test

File: src/tests/exchange/test_liquidation_handler.cairo [hunk 1]
```diff
@@ -34,6 +50,42 @@ fn test_exec_liquidation_true() {
...
+    let mut position: Position = create_new_position(
+        key, account, market, collateral_token, is_long, position_no: 1
+    );
+
+    data_store.set_position(key, position);
+
+    let mut state_base: BaseOrderHandler::ContractState =
+        BaseOrderHandler::unsafe_new_contract_state(); //retrieve BaseOrderHandler state
+
+    let key: felt252 = create_liquidation_order(
```
Comment (lambda-0x) [inline src/tests/exchange/test_liquidation_handler.cairo:75]: to test `create_liquidation_order` which creates an order and add it to `data_store` i have added this test. but it failing with cryptic error message (ig its coming from call to `unsafe_new_contract_state`)

```
Collected 1 test(s) and 2 test file(s)
Running 1 test(s) from satoru package
[FAIL] satoru::tests::exchange::test_liquidation_handler::test_create_liquidation_order

Failure data:
    Got an exception while executing a hint:
    Bad syscall_ptr; expected: Relocatable { segment_index: 10, offset: 0 }, got: Relocatable { segment_index: 10, offset: 60 }.

Running 0 test(s) from tests/utils/test_pricing_utils.cairo
Tests: 0 passed, 1 failed, 0 skipped

Failures:
    satoru::tests::exchange::test_liquidation_handler::test_create_liquidation_order
```

but i needed to call it to get `event_emitter` required by `create_liquidation_order` function. Any other way to get `event_emitter`?

Comment (lambda-0x) [reply src/tests/exchange/test_liquidation_handler.cairo:75]: cc: @Sk8erboi84

Comment (sparqet) [reply src/tests/exchange/test_liquidation_handler.cairo:75]: you have passed wrong arguments in the function, you should deploy an event_emitter contract in setup() and pass it to create_liquidation_order. do the same for data_store, without calling base_order_handler

Comment (lambda-0x) [reply src/tests/exchange/test_liquidation_handler.cairo:75]: `setup` already deploys data_store, so i think i will only need to add `event_emitter`. should i make this change in `setup` function (requiring to update every call to this function) or create a new function called `setup_event_emitter`

Comment (lambda-0x) [reply src/tests/exchange/test_liquidation_handler.cairo:75]: got it, thanks!

Comment (sparqet) [reply src/tests/exchange/test_liquidation_handler.cairo:75]: You need to update setup() and pass data_store returned by setup() to create_liquidation fonction

Comment (lambda-0x) [reply src/tests/exchange/test_liquidation_handler.cairo]: yeah realized that just after posting previous comment, its fixed.

**Comment by lambda-0x** [comment]
still getting that weird error message

```
Collected 1 test(s) and 2 test file(s)
Running 1 test(s) from satoru package
[FAIL] satoru::tests::exchange::test_liquidation_handler::test_create_liquidation_order

Failure data:
    Got an exception while executing a hint:
    Bad syscall_ptr; expected: Relocatable { segment_index: 10, offset: 0 }, got: Relocatable { segment_index: 10, offset: 90 }.

Running 0 test(s) from tests/utils/test_pricing_utils.cairo
Tests: 0 passed, 1 failed, 0 skipped

Failures:
    satoru::tests::exchange::test_liquidation_handler::test_create_liquidation_order
```

**Comment by lambda-0x** [comment]
upstream issue: https://github.com/foundry-rs/starknet-foundry/issues/659

**Comment by lambda-0x** [comment]
(commented out the failing test, should be uncommented after upstream issue is fixed)

File: tests/exchange/test_liquidation_handler.cairo [hunk 2]
```diff
@@ -14,16 +15,36 @@ use satoru::role::role;
...
+use satoru::position::position::Position;
+use satoru::liquidation::liquidation_utils::create_liquidation_order;
+use satoru::exchange::base_order_handler::{IBaseOrderHandler, BaseOrderHandler};
+use satoru::exchange::base_order_handler::BaseOrderHandler::{
+    event_emitter::InternalContractMemberStateTrait, data_store::InternalContractMemberStateImpl
+};
+use satoru::event::event_emitter::{IEventEmitterDispatcher};
 
 #[test]
 fn test_exec_liquidation_true() {
```
Comment (sparqet) [inline tests/exchange/test_liquidation_handler.cairo:27]: rename test with the current format given_when_then

**Review by sparqet — APPROVED** [review]

**Pull request merged**